### ISSUE TITLE
Modelica_Noise Library for inclusion in Modelica Standard Library

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1662]
**Reported by otter on 19 Feb 2015 11:25 UTC**
The Modelica_Noise library from DLR-SR is available under:
https://github.com/DLR-SR/Noise/tree/MSL

Click on the right side on "Download zip" to get the library. Inspect only "Modelica_Noise 1.0-Alpha.1" and ignore the "Noise 1.0-Alpha.1" (which contains components/models not yet ready for MSL).

It is suggested to include this library into the Modelica Standard Library.

The library provides basically 2 blocks to generated reproducible noise:
- Modelica_Noise.Blocks.Noise.TimeBasedNoise (this block computes always the same (random) output y at the same time instant for the same model instance)
- Modelica_Noise.Blocks.Noise.SignalBasedNoise (this block computes always the same (random) output y at the same value of the input signal for the same model instance).

The noise generation is based on the xorshift suite of random number generators from Sebastiano Vigna (from 2014; http://xorshift.di.unimi.it/). These random number generators have excellent statistical properties and have a number of advantages over previous generators. An overview of the features and advantages is provided here:
  Modelica_Noise.Math.Random.Generators

The Modelica_Noise library is a completely restructured, and largely newly implemented version of the Noise package from ticket #1514. All comments of this ticket have been taken into account.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1662

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 19 Feb 2015 12:21 UTC**
Additional information:

The Modelica_Noise library has the following advantages:
- Produces either event-based noise (= traditional) or event-free noise (new). The traditional signal noise implementation has the disadvantage that simulation time slows down drastically (say factor 10-100). The event-free option has no essential reduction of efficiency, if the relative tolerance of the integrator is not too tight (say >= 1e-6).
- Every noise block has its own random state and is initialized with a different seed, by default a hash value of its instance name; a suggestion from Martin Sjölund. This gives reproducible signal noise, as it is important for control applications (for example when changing a parameter or a component of a control system, or adding a new control component, and re-simulating, it is important that the signal noise does not change, because otherwise it is hard to determine whether the changed behavior is due to the changes of the control system or due to the different signal noise). 
- The previous version of the library used quite sophisticated constructs (function objects in replaceable functions). This new version uses only replaceable functions and packages (so similar as Modelica.Media, but much simpler). Therefore, the hope is that all the Modelica tools can support it.
- The Modelica_Noise library is not a new sub-library, but it basically introduces two new blocks in the Modelica.Blocks library and several new utility functions under Modelica.Math and Modelica.Utilities. However, since the blocks use "replaceable" and require additional C-code, a discussion in the group and with the tool vendors is needed before inclusion.

Before inclusion into the MSL, the following (minor) issues need to be solved/agreed on with the tool vendors:
- The current implementation includes the C-code with the include annotation. Hopefully some tool vendors can help to change this, so that it works in the same way as the other C-code in MSL (similarly as the table blocks);
- MSL 3.2.x only uses Modelica 3.2 language elements. The Modelica_Noise library uses two Modelica 3.3 language elements: (a) getInstanceName() and (b) "impure" keyword. If desired, the "impure" keyword can be removed (the advantage is that it clearly marks the impure functions). The "getInstanceName()" is essential, because it is used to compute an automatic seed whenever a noise block is dragged.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 19 Feb 2015 12:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 19 Feb 2015 15:59 UTC**
If Modelica 3.3 language elements are used it cannot be added to an MSL 3.2.x. This would break the entire release process otherwise.

See #1478 for a discussion on the Impure annotation. Martin S. suggested to make an MSL 3.3 the next release. This would also solve the issue with getInstanceName().

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 22 Jun 2015 14:28 UTC**
A new version of the Modelica_Noise library has been committed to github. It is available under: https://github.com/DLR-SR/Noise/tree/master

If you do not use git, just download: https://github.com/DLR-SR/Noise/archive/master.zip unzip this file and open Noise-master/Modelica_Noise 1.0-Beta.1.

The following changes are made with respect to the previous Alpha.1 version:
- TimeBasedNoise block replaced by GenericNoise block (uses the traditional technique to periodically generate a time event, draw a random number and keep this value constant until the next time event).
- Removed SignalBasedNoise block.
- Added BandLimitedWhiteNoise block.
- Only Modelica 3.2 language elements are used.
- Improved the examples
- Adapted and improved the documentation.

According to our understanding, all reported issues are fixed and remarks taken into account. Please, re-evaluate the library.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 14 Jul 2015 06:56 UTC**
Christoph Clauß added an [attachment:EvaluationModelicaLibraryCandidate_Modelica_Noise.pdf​]

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jul 2015 20:21 UTC**
Replying to [comment:5 dietmarw]:

> Christoph Clauss added an [attachment:EvaluationModelicaLibraryCandidate_Modelica_Noise.pdf]
> 

Thanks for the evaluation. Here is a respond to the comments:

Note, ratings are from 1 to 5, where 5 is the best rating

**Answers to review of Christoph Clauss (the library shall be included in the MSL: rating = 5)**

_Comment 1: Users'Guide (rating: 1)_
Respond: There is no User's Guide because this is a collection of models. In the package description of Blocks.Noise, the usage of the noise blocks is explained (corresponds to a User's Guide).

_Comment 2: Since TruncatedDistributions are distributions I'd expect them to be located under Distributions. I suggest the truncated versions to locate in such a way:_

```
Distributions
  Normal
  TruncatedNormal ...
```
Respond: You are right, this might be a possibility. The drawback is that package Distributions has then 2 sub-packages only (so quite overkill for one package), and the full names are a bit strange (Modelica.Math.Distributions sounds better as Modelica.Math.Distributions.Normal; Modelica.Math.TruncatedDistributions sounds better as Modelica.Math.Distributions.TruncatedNormal). So, we would like to keep the current structuring

```
Math
  ...
  Distributions
  TruncatedDistributions

```
_Comment 3: Documentation should be improved:_

 * _If a name is mentioned, a source should follow. E.g. "suite developed in 2014 by Sebastiano Vigna" in the package Noise._
    Respond: You are right. This will be fixed (a link to the random package and to the web site of Vigna will be added).

 * _Some functions are too shortly documented, e.g. hashString In the Test examples the sense of the examples should be explained in slightly more detail._
    Respond: Hm, we recognized that no test for hashString is any longer present (also not in Modelica_Noise.ToModelicaTest), due to removal from automatic seed calculation in the Noise blocks. We will add one example in Modelica_Noise.ToModelicaTest, where this function shall be called and explain the usage.
   
 * _Some examples for using the random generation for MonteCarlo simulations with randomly assigned parameter values during repeated simulations runs are missing._
   Respond: This cannot be described in Modelica (because parameters are fixed for one simulation run and it is not possible to define in Modelica several simulation runs). Typically, a scripting language is needed to define MonteCarlo simulations where parameters are changed statistically in every simulation run. We will add a short explanation in the Blocks.Noise package.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jul 2015 20:51 UTC**
Replying to

> Attachment https://trac.modelica.org/Modelica/attachment/ticket/1662/EvaluationModelicaLibraryCandidate_Modelica_Noise.2.pdf added
> 
> Hans' evaluation (several comments inside)

Thanks for the evaluation. Here is a respond to the comments:

Note, ratings are from 1 to 5, where 5 is the best rating

**Answers to review of Hans Olsson (the library shall be included in the MSL: rating = 4)**

_Comment 1: To me it seems that a better alternative is to place the entire Modelica_Noise.Blocks.Noise in Modelica.Blocks.Sources.Noise (it is similar to other sources, but still a separate package since there are no other noise sources)._
Respond: We agree that the two noise blocks have only output signals and therefore the natural location would be under Blocks.Sources. On the other hand, there is also the GlobalSeed component, that is not a Sources component. We also had in the past noise blocks with inputs, but removed them for this version. We plan to include at least one Noise block in the future with one input and one output (the input is the ideally measured signal, and the output is the input with added noise) to simplify the usage for measurement signals. It seems then more natural to place them under package Blocks.Noise, but not under Blocks.Sources.Noise. So, we have the feeling it is more future proof, to keep the blocks under Blocks.Noise.

_Comment 2: Place Modelica_Noise.Blocks.Statistics in Modelica.Blocks.Math (which already contain Mean, RectifiedMean, Max, etc.). A later reorganization would have all the Statistics blocks in a sub-package of Math._
Respond: We agree and will move the blocks from Statistics to Math.

_Comment 3: The Density-block does not belong in the Statistics library; the other estimates statistical properties of a signal (and it would be good to have some way of estimating density as well), but the Density block is more like an Example showing what the Density function looks like, and should thus be placed in the Example library_
Respond: We agree and will move the block.

_Comment 4: The library shall be included in the MSL (rating 4): The only counter-argument is the growing size of MSL and possibly that it should be added to MSL 3.3 not to 3.2 (see below)._
Respond: Random number generation and noise blocks are urgently missing for several application areas and it is hard  to provide this in a portable way so that such blocks work in all Modelica tools. It is then best to include this base functionality in the MSL (where vendors give support so that the models run in their tool). We agree that including this for MSL 3.3.x instead of MSL 3.2.x would be better, because then the language elements "impure" and "getInstanceName" could be used. However, since currently only Dymola supports Modelica 3.3, it seems overkill to require support of Modelica 3.3 language elements. Therefore, we think that the Modelica 3.2 language elements restriction must be accepted, although this has slight drawbacks.

_Comment 5: Modelica_Noise is compatible with Modelica 3.3, not with 3.2 due to the use of "impure" in e.g. Modelica_Noise.Utilities.System.getTime. The evaluation does not indicate which Modelica version is meant. Either that keyword should be removed, or the library can be rewritten to take advantage of getInstance in 3.3 to make individual sources independent._
Respond: The library is planned for Modelica 3.2 version. The usage of the "impure" keyword is an oversight. We checked, and the keyword "impure" is used for "Modelica_Noise.Math.Random.Utilities.automaticGlobalSeed", "Modelica_Noise.Math.Random.Utilities.automaticLocalSeed", "Modelica_Noise.Utilities.System.getPid", "Modelica_Noise.Utilities.System.getTime". We removed the "impure" keyword and committed this to github (these functions are only used in when-clauses and are external functions, and therefore a tool should not assume that these functions are "pure").

_Comment 6: The description of truncated distributions is not clear (it is not clear why they are useful); and it is also confusing that the Wikipedia entry states that there are two alternatives - and one has to search to determine which one is used._
Respond: The linked Wikipedia article only describes one truncated distribution, and discusses one possible other definition, but explains shortly why this other definition is not meaningful. We will make this clearer in the documentation of package TruncatedDistribution. The main reason to introduce truncated distributions is to make the modeling of measurement noise easier, in order to limit the band in which the noise can occur (in reality the noise is not unlimited). For example, if a sensor is used and the sensor signal inspected for an output of 0 Volt and the measured signal is +/- 0.1 Volt, then one has a quite good knowledge about the band width of the noise (this can be easily inspected for nearly every sensor). The sensor signal will be the input to an Analog-Digital block and this block limits the signal, say to +/- 5 Volt. Typically, the user would like to model noise with a band of +/- 0.1 Volt, and often uses a normal distribution. But a normal distribution is not limited and for a small sample time and a long simulation there might be some sample time instants where the noise values of the normal signal is outside the +/- 0.1 Volt range. For some sensor types this is completely unrealistic (e.g. an angle sensor might measure +/- 0.1 rad, but the sensor will never add, say one revolution (6.28 rad) to it. However, the noise model with a pure normal distribution could give such a value. If a user would like to guarantee (and not to hope), that the modeled noise is always between +/- 0.1 Volt, then there are two main possibilities: (a) The noise is computed and the result is then limited to +/- 0.1 Volt, or (b) the normal distribution is slightly modified, so that it is within the band of +/- 0.1 Volt. Approach (a) is a brute force method that changes the statistical properties of the signal in an unknown way. Approach (b) is a "clean" description. The Noise block gives the user the freedom to choose: Either compute a normal (unlimited) noise, or a truncated normal noise (truncated distribution). We will improve the documentation of the Noise block and the TruncatedDistribution with this explanation.

_Comment 7: Some minor issues: Modelica_Noise.Blocks.Statistics uses b-tag instead of strong-tag._
Respond: In the whole MSL the b-tag is usually used and not the strong-tag. So, we keep the b-tag (which is used for all bold-face fonts in Modelica_Noise). If this is not desired, all `<b>` tags in MSL should be changed to `<strong>` tags at one point in time.

_Comment 8: Would really prefer if the Modelica_Noise.Blocks.Noise.GlobalSeed had useAutomaticSeed=true and getInstanceName-variant used to ensure that the noise is non-deterministic (or if not possibly just have a deafult of non-repeatable noise-generators). An alternative (which is less good) is to only have start-values for these parameters forcing the user to specify them manually. Basically, to me using the same automatic seed for multiple runs and generators is problematic in the same way as using "R=1" for a resistor. Thus in the spirit of the rest of the library it should be avoided._
Respond: For most applications in the control community the major application of the noise blocks is to use them in controlled plant models. For the design and assessment of controllers it is absolutely essential that the noise is reproducible, so that in every simulation run exactly the same noise is used (so GlobalSeed.useAutomaticSeed=false). The only application where in every simulation run a different noise shall be utilized are Monte-Carlo simulations. Since the majority of applications (at least seen from a control engineering perspective) are from the first kind, we prefer a default of GlobalSeed.useAutomaticSeed=false, in order that users are not surprised and make errors in the most common case. Note, for Monte-Carlo simulations typically one would have another program (say a script), and this program would take care that the random numbers are different for every new simulation run.We also would prefer to use the getInstanceName() function. However, in order to be Modelica 3.2 compliant we cannot. Instead, the seed in the noise blocks has a default of "useAutomaticLocalSeed=true" and the seed is computed by globalSeed.randomInteger() (this is the impure random number generator Modelica_Noise.Blocks.Noise.GlobalSeed.random (xorshift1024\*) where the result is mapped to an integer with Modelica_Noise.Math.Random.Utilities.impureRandomInteger). Since the Noise blocks use another random number generator (Xorshift128+), the random number sequences of different noise blocks are not correlated by default. Therefore, just dragging the noise blocks will generate different noise signals.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 16 Jul 2015 11:07 UTC**
Replying to [comment:7 otter]:
> Replying to
> 
> > Attachment https://trac.modelica.org/Modelica/attachment/ticket/1662/EvaluationModelicaLibraryCandidate_Modelica_Noise.2.pdf added
> > 
> > Hans' evaluation (several comments inside)
> 
> Thanks for the evaluation. Here is a respond to the comments:
> 
> Note, ratings are from 1 to 5, where 5 is the best rating
> 
> **Answers to review of Hans Olsson (the library shall be included in the MSL: rating = 4)**
> 
> _Comment 1: To me it seems that a better alternative is to place the entire Modelica_Noise.Blocks.Noise in Modelica.Blocks.Sources.Noise (it is similar to other sources, but still a separate package since there are no other noise sources)._
> Respond: We agree that the two noise blocks have only output signals and therefore the natural location would be under Blocks.Sources. On the other hand, there is also the GlobalSeed component, that is not a Sources component. We also had in the past noise blocks with inputs, but removed them for this version. We plan to include at least one Noise block in the future with one input and one output (the input is the ideally measured signal, and the output is the input with added noise) to simplify the usage for measurement signals. It seems then more natural to place them under package Blocks.Noise, but not under Blocks.Sources.Noise. So, we have the feeling it is more future proof, to keep the blocks under Blocks.Noise.

Ok, I can see that argument.

> _Comment 4: The library shall be included in the MSL (rating 4): The only counter-argument is the growing size of MSL and possibly that it should be added to MSL 3.3 not to 3.2 (see below).
> _       Respond: Random number generation and noise blocks are urgently missing for several application areas and it is hard  to provide this in a portable way so that such blocks work in all Modelica tools. It is then best to include this base functionality in the MSL (where vendors give support so that the models run in their tool). We agree that including this for MSL 3.3.x instead of MSL 3.2.x would be better, because then the language elements "impure" and "getInstanceName" could be used. However, since currently only Dymola supports Modelica 3.3, it seems overkill to require support of Modelica 3.3 language elements. Therefore, we think that the Modelica 3.2 language elements restriction must be accepted, although this has slight drawbacks.

I understand, but the problem is then if we can ensure that it will be future-proof once we release MSL 3.3 - i.e. how we interpret backward compatibility to make it possible to add the getInstanceName etc. Assuming that has been analyzed it should be ok to add it to 3.2.

> _Comment 7: Some minor issues: Modelica_Noise.Blocks.Statistics uses b-tag instead of strong-tag._
> Respond: In the whole MSL the b-tag is usually used and not the strong-tag. So, we keep the b-tag (which is used for all bold-face fonts in Modelica_Noise). If this is not desired, all `<b>` tags in MSL should be changed to `<strong>` tags at one point in time.

I have opened another ticket about that issue #1747 - I agree it is not a problem with this library. However, I only followed the evaluation form.
Note that `<b>` tags should NOT blindly be replaced by `<strong>` tags.

> _Comment 8: Would really prefer if the Modelica_Noise.Blocks.Noise.GlobalSeed had useAutomaticSeed=true and getInstanceName-variant used to ensure that the noise is non-deterministic (or if not possibly just have a deafult of non-repeatable noise-generators). An alternative (which is less good) is to only have start-values for these parameters forcing the user to specify them manually. Basically, to me using the same automatic seed for multiple runs and generators is problematic in the same way as using "R=1" for a resistor. Thus in the spirit of the rest of the library it should be avoided._
> Respond: For most applications in the control community the major application of the noise blocks is to use them in controlled plant models. For the design and assessment of controllers it is absolutely essential that the noise is reproducible, so that in every simulation run exactly the same noise is used (so GlobalSeed.useAutomaticSeed=false). The only application where in every simulation run a different noise shall be utilized are Monte-Carlo simulations. Since the majority of applications (at least seen from a control engineering perspective) are from the first kind, we prefer a default of GlobalSeed.useAutomaticSeed=false, in order that users are not surprised and make errors in the most common case. Note, for Monte-Carlo simulations typically one would have another program (say a script), and this program would take care that the random numbers are different for every new simulation run.We also would prefer to use the getInstanceName() function. However, in order to be Modelica 3.2 compliant we cannot. Instead, the seed in the noise blocks has a default of "useAutomaticLocalSeed=true" and the seed is computed by globalSeed.randomInteger() (this is the impure random number generator Modelica_Noise.Blocks.Noise.GlobalSeed.random (xorshift1024\*) where the result is mapped to an integer with Modelica_Noise.Math.Random.Utilities.impureRandomInteger). Since the Noise blocks use another random number generator (Xorshift128+), the random number sequences of different noise blocks are not correlated by default. Therefore, just dragging the noise blocks will generate different noise signals.

I agree that having it reproducible is important (and we need getInstanceName to make it reproducible with regards to changes in the model).

But that is subtly different from having it always generating the same result, and my fear is that users will not recognize this and I feel that as default producing different result is the safe choice.

For reproducibility if you set 'useAutomaticSeed=true' you can reproduce the simulation by extracting 'globalSeed.seed' (and changing to useAutomaticSeed=false).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 17 Jul 2015 07:53 UTC**
Replying to [comment:6 otter]:
> Replying to [comment:5 dietmarw]:
> 
> > Christoph Clauss added an [attachment:EvaluationModelicaLibraryCandidate_Modelica_Noise.pdf]
> > 
> 
> Thanks for the evaluation. Here is a respond to the comments:
> 
> Note, ratings are from 1 to 5, where 5 is the best rating
> 
> **Answers to review of Christoph Clauss (the library shall be included in the MSL: rating = 5)**
> 
> _Comment 1: Users'Guide (rating: 1)_
> Respond: There is no User's Guide because this is a collection of models. In the package description of Blocks.Noise, the usage of the noise blocks is explained (corresponds to a User's Guide).
> 
> _Comment 2: Since TruncatedDistributions are distributions I'd expect them to be located under Distributions. I suggest the truncated versions to locate in such a way:_
> 
> ```
> Distributions
>   Normal
>   TruncatedNormal ...
> ```
> Respond: You are right, this might be a possibility. The drawback is that package Distributions has then 2 sub-packages only (so quite overkill for one package), and the full names are a bit strange (Modelica.Math.Distributions sounds better as Modelica.Math.Distributions.Normal; Modelica.Math.TruncatedDistributions sounds better as Modelica.Math.Distributions.TruncatedNormal). So, we would like to keep the current structuring
> 
> ```
> Math
>   ...
>   Distributions
>   TruncatedDistributions
> 
> ```

Sorry, I misunderstood your proposal. It is a good idea to move the contents of TruncatedDistributions to Distributions (so have packages Uniform, Normal, TruncatedNormal, Weibull, TruncatedWeibull under package Distributions).

> _Comment 3: Documentation should be improved:_
> 
>  \* _If a name is mentioned, a source should follow. E.g. "suite developed in 2014 by Sebastiano Vigna" in the package Noise._
   Respond: You are right. This will be fixed (a link to the random package and to the web site of Vigna will be added).
> 
>  \* _Some functions are too shortly documented, e.g. hashString In the Test examples the sense of the examples should be explained in slightly more detail._
   Respond: Hm, we recognized that no test for hashString is any longer present (also not in Modelica_Noise.ToModelicaTest), due to removal from automatic seed calculation in the Noise blocks. We will add one example in Modelica_Noise.ToModelicaTest, where this function shall be called and explain the usage.
>   
>  \* _Some examples for using the random generation for MonteCarlo simulations with randomly assigned parameter values during repeated simulations runs are missing._
   Respond: This cannot be described in Modelica (because parameters are fixed for one simulation run and it is not possible to define in Modelica several simulation runs). Typically, a scripting language is needed to define MonteCarlo simulations where parameters are changed statistically in every simulation run. We will add a short explanation in the Blocks.Noise package.

Improved the Modelica_Noise library according to the suggestions above and committed to github.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 17 Jul 2015 08:23 UTC**
Improved Modelica_Noise according to the discussion in comment https://trac.modelica.org/Modelica/ticket/1662#comment:7 (suggestions from Hans).

Additional comments:

 * Included a comment in Blocks.Noise that automaticLocalSeed will be changed to "hash value of full component name", once it is allowed to use Modelica Language 3.3 elements in the MSL.
 * Experimented to have "GlobalSeed.useAutomaticSeed(start=false)" in order to force to provide a value here (otherwise a warning is printed when simulation starts, or an error if pedantic). However, the visual appearance in Dymola is exactly the same as with "GlobalSeed.useAutomaticSeed(start=false)" because a check box is used. Since without a checkbox the menu does not look so nice, I kept the current default false, and improved the icon by adding "fixedSeed = %fixedSeed". Therefore, if useAutomaticSeed=false, it is very visible that a fixed seed is used (fixedSeed is not displayed if useAutomaticSeed=true).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 17 Jul 2015 08:39 UTC**
Francesco, thanks for the evaluation. Your comments have been incorporated and committed to github:

_Models in the Statistics block extend Modelica.Blocks.Interfaces.BlockIcon, they should extend Modelica.Blocks.Icons.Block instead_
This has been fixed.

_The last example DrydenContinuousTurbulence lacks the inner
globalSeed object, please add it explicitly to the codeThanks for the evaluation._
This has been fixed.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 Jul 2015 20:57 UTC**
Replying to [comment:1 otter]:
> Additional information:
> Before inclusion into the MSL, the following (minor) issues need to be solved/agreed on with the tool vendors:
> - The current implementation includes the C-code with the include annotation. Hopefully some tool vendors can help to change this, so that it works in the same way as the other C-code in MSL (similarly as the table blocks);

Indeed, ModelicaRandom.c needs to be converted to an include file and an external library for usage with SimulationX.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 5 Oct 2015 13:06 UTC**
Added implementation of ModelicaStrings_hashString, ModelicaInternal_getpid and ModelicaInternal_getTime to external C sources by 61763832d4c6ce46403919c11a5cd5ff396d6735.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 5 Oct 2015 14:56 UTC**
ModelicaRandom.c (modified) was added by f329842ea17b12b9602015d45743d032aecdb5a3. This file should be compiled within ModelicaExternalC library provided by the tool vendors.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 21 Oct 2015 10:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 10 Dec 2015 07:45 UTC**
Corresponding changesets: dc25f7c87190a74b32b948032d4863c7ab713165, 205941ff6e4cdb68d5d6c958d72a7b49dbbfc28a, 189484fb8ecdab302927e322e0143a61f317b598

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 10 Dec 2015 08:33 UTC**
Added all functions from Modelica_Noise (dc25f7c87190a74b32b948032d4863c7ab713165, 205941ff6e4cdb68d5d6c958d72a7b49dbbfc28a, 189484fb8ecdab302927e322e0143a61f317b598, 3fc1bd289f445cc9886a25cbb9634ee8c4ec17a7, 883e026a0e2e45570f5880c72783109b5e108b80, 06bcb3036868c2f9aa354ca865b04140461e020d).

The functions depend on C-functions stored in Resources/C-Sources/ModelicaRandom.C, ModelicaInternal.C. Therefore, noise models can only be simulated if the ModelicaExternalC library is newly build with these functions. The newly built library is at least available in Dymola 2016 FD01 and in SimulationX (https://github.com/ITI-GmbH/ModelicaExternalC).

However, the Modelica.Math.Random.Utilities.automaticGlobalSeed C-function was just added and is not yet included in any of the tools. For this reason the actual implementation of "automaticGlobalSeed" is temporarly not as "external C" function but it returns only a fixed Integer. Once tools have built a new ModelicaExternalC library, this implementation must be adapted.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 11 Dec 2015 12:12 UTC**
Included Modelica_Noise library in MSL.
If you find issues, open a separate ticket.

**Comment by modelica-trac-importer** [comment]
**Changelog modified by otter on 11 Dec 2015 12:12 UTC**
Included functions and blocks for random number generation with specified distributions, and to model different types of (measurement) noise

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 14 Dec 2015 16:44 UTC**
I have problem with the library having unused inputs in external functions. OpenModelica will detect this and the artificial dependencies which are introduced will be ignored.

Also, the library uses impure functions a lot, but they are not marked as such. OpenModelica will optimize this because it knows that the specification says functions are pure.

I really would like to avoid having to patch this entire part of the library because it did not use language constructs that were designed in order to handle the package.
```
[Modelica trunk/Math/Random.mo:906:7-906:75:writable] Warning: Unused input [Modelica trunk/Math/Random.mo:906:7-906:75:writable] Warning: Unused input variable stateSize in function .Modelica.Math.Random.Generators.Xorshift64star.random.
```

```
assert            | debug   | Function impureRandom not initialized with function initializeImpureRandom
assert            | info    | simulation terminated by an assertion at initialization
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Dec 2015 07:17 UTC**
Fixed "unused input stateSize" in 842e15c62e1dadde6ca450b34eee4d318eddaa52

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Dec 2015 07:37 UTC**
Fixed "Function impureRandom not initialized with function initializeImpureRandom" in d59ba164b7437351413ad9c98bf957e1b42134dc and 423005fea3402c0227adfdb90f3d324df4df6e4f by rewriting the initialization in such a way that a tool is forced to call the impure random number generation initializiation during initialization (and not during translation of the model).

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 Dec 2015 10:28 UTC**
Reopen such that we do not forget to remove the temporary hack from source:Modelica/trunk/Modelica/Math/Random.mo#L979

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 Dec 2015 11:48 UTC**
Current trunk gives translation errors that can be avoided by the attachment:T1662.patch

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Dec 2015 13:07 UTC**
Applied patch on random.mo; The Noise.mo on svn repository was already identical to the Noise.mo provided by the patch ("Update" forgotten?)

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 Dec 2015 13:11 UTC**
Replying to [comment:24 otter]:
> Applied patch on random.mo; The Noise.mo on svn repository was already identical to the Noise.mo provided by the patch ("Update" forgotten?)

No, it's not. Please patch also Noise.mo by attachment:T1662.patch​

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Dec 2015 13:30 UTC**
Hm. It seems my changes have not been committed. The patch was not correct. I have now committed the correct code (an alternative to "initial equation" is "when initial()").
I recognized that I also have to slightly rewrite impureRandom() and impureRandomInteger() to guarantee that in all cases a tool cannot optimize the call in a "wrong" way.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 23 Dec 2015 06:47 UTC**
In OpenModelica it was sufficient to rewrite the initialization in order that the random number generators work. Therefore, I close this ticket.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 23 Dec 2015 12:52 UTC**
Replying to [comment:22 beutlich]:
> Reopen such that we do not forget to remove the temporary hack from source:Modelica/trunk/Modelica/Math/Random.mo#L979

Reopen again as long as that hack is not removed. This way we can keep track of it.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 12 Jan 2016 12:41 UTC**
Fixed in 94d166b598a8a203fab80939a337b0dea2068fbb
